src/rnnlm/rnnlm-simple-looped.cc

```diff
@@ -0,0 +1,200 @@
+    CuVector<BaseFloat> log_probs(word_embedding_mat.NumRows());
+
+    log_probs.AddMatVec(1.0, word_embedding_mat, kTrans, hidden, 0.0);
+    log_probs.ApplySoftMax();
+    log_probs.ApplyLog();
+    log_prob = log_probs(word_index);
+  } else {
+    log_prob = VecVec(hidden, word_embedding_mat.Row(word_index));
+  }
+  return log_prob;
+}
+
+CuVector<BaseFloat>* RnnlmSimpleLooped::GetOutput(int32 frame) {
+  KALDI_ASSERT(frame >= 0 && frame < feats_.NumRows());
+  if (frame >= current_log_post_offset_ + current_nnet_output_.NumRows())
+    AdvanceChunk();
+
+//  int32 embedding_dim = info_.word_embedding_mat.NumCols();
+//  int32 num_words = info_.word_embedding_mat.NumRows();
+
+
+  CuMatrix<BaseFloat> current_nnet_output_gpu;
+  current_nnet_output_gpu.Swap(&current_nnet_output_);
+  const CuSubVector<BaseFloat> hidden(current_nnet_output_gpu,
+                                      frame - current_log_post_offset_);
+  return new CuVector<BaseFloat>(hidden);
+//
+//  // swap the pointer back so that this function can be called multiple times
+//  // with the same returned value before taking next new feats
+//  current_nnet_output_.Swap(&current_nnet_output_gpu);
+//  return log_prob;
+}
+
+void RnnlmSimpleLooped::AdvanceChunk() {
+  int32 begin_input_frame, end_input_frame;
+  begin_input_frame = -info_.frames_left_context;
+  // note: end is last plus one.
+  end_input_frame = info_.frames_per_chunk + info_.frames_right_context;
+  // currently there is no left/right context and frames_per_chunk == 1
+  KALDI_ASSERT(begin_input_frame == 0 && end_input_frame == 1);
+
+  SparseMatrix<BaseFloat> feats_chunk(end_input_frame - begin_input_frame,
+                                      feats_.NumCols());
+  int32 num_features = feats_.NumRows();
+  for (int32 r = begin_input_frame; r < end_input_frame; r++) {
+    int32 input_frame = r;
+    if (input_frame < 0) input_frame = 0;
+    if (input_frame >= num_features) input_frame = num_features - 1;
+    feats_chunk.SetRow(r - begin_input_frame, feats_.Row(input_frame));
+  }
+
+  CuMatrix<BaseFloat> input_embeddings(1, info_.word_embedding_mat.NumCols());
+  int32 word_index = feats_chunk.Row(0).GetElement(0).first;
+  input_embeddings.RowRange(0, 1).AddMat(1.0, info_.word_embedding_mat.RowRange(word_index, 1), kNoTrans);
+  computer_.AcceptInput("input", &input_embeddings);
+
+  computer_.Run();
+
+  {
+    // Note: here GetOutput() is used instead of GetOutputDestructive(), since
+    // here we have recurrence that goes directly from the output, and the call
+    // to GetOutputDestructive() would cause a crash on the next chunk.
+    CuMatrix<BaseFloat> output(computer_.GetOutput("output"));
+
+    current_nnet_output_.Resize(0, 0);
+    current_nnet_output_.Swap(&output);
+  }
+  KALDI_ASSERT(current_nnet_output_.NumRows() == info_.frames_per_chunk &&
+               current_nnet_output_.NumCols() == info_.nnet_output_dim);
+
+  current_log_post_offset_ = 0;
```

Comment:
This doesn't make sense.  If the variable is always zero, why have the variable at all?